File: lib/web_ui/lib/src/engine/compositor/color_filter.dart
// Copyright 2013 The Flutter Authors. All rights reserved.
// Use of this source code is governed by a BSD-style license that can be
// found in the LICENSE file.


part of engine;

/// A [ui.ColorFilter] backed by Skia's [CkColorFilter].
class CkColorFilter extends ResurrectableSkiaObject {
  final EngineColorFilter _engineFilter;

  CkColorFilter.mode(EngineColorFilter filter) : _engineFilter = filter;

  CkColorFilter.matrix(EngineColorFilter filter) : _engineFilter = filter;

  CkColorFilter.linearToSrgbGamma(EngineColorFilter filter)
      : _engineFilter = filter;

  CkColorFilter.srgbToLinearGamma(EngineColorFilter filter)
      : _engineFilter = filter;

  js.JsObject _createSkiaObjectFromFilter() {
    switch (_engineFilter._type) {
      case EngineColorFilter._TypeMode:
        setSharedSkColor1(_engineFilter._color!);
        return canvasKit['SkColorFilter'].callMethod('MakeBlend', <dynamic>[
          sharedSkColor1,
          makeSkBlendMode(_engineFilter._blendMode),
        ]);
      case EngineColorFilter._TypeMatrix:
        final js.JsArray<double> colorMatrix = js.JsArray<double>();
        colorMatrix.length = 20;
        for (int i = 0; i < 20; i++) {
          colorMatrix[i] = _engineFilter._matrix![i];
        }
        return canvasKit['SkColorFilter']
            .callMethod('MakeMatrix', <js.JsArray>[colorMatrix]);
      case EngineColorFilter._TypeLinearToSrgbGamma:
        return canvasKit['SkColorFilter'].callMethod('MakeLinearToSRGBGamma');
      case EngineColorFilter._TypeSrgbToLinearGamma:
        return canvasKit['SkColorFilter'].callMethod('MakeSRGBToLinearGamma');
      default:
        throw StateError(
            'Unknown mode ${_engineFilter._type} for ColorFilter.');
    }
  }

  @override
  js.JsObject createDefault() {
    return _createSkiaObjectFromFilter();
  }

  @override
  js.JsObject resurrect() {
    return _createSkiaObjectFromFilter();
  }
}

File: engine/src/flutter/impeller/toolkit/interop/color_filter.cc
// Copyright 2013 The Flutter Authors. All rights reserved.
// Use of this source code is governed by a BSD-style license that can be
// found in the LICENSE file.

#include "impeller/toolkit/interop/color_filter.h"

namespace impeller::interop {

ScopedObject<ColorFilter> ColorFilter::MakeBlend(Color color, BlendMode mode) {
  auto filter = flutter::DlBlendColorFilter::Make(ToDisplayListType(color),
                                                  ToDisplayListType(mode));
  if (!filter) {
    return nullptr;
  }
  return Create<ColorFilter>(std::move(filter));
}

ScopedObject<ColorFilter> ColorFilter::MakeMatrix(const float matrix[20]) {
  auto filter = flutter::DlMatrixColorFilter::Make(matrix);
  if (!filter) {
    return nullptr;
  }
  return Create<ColorFilter>(std::move(filter));
}

ColorFilter::ColorFilter(std::shared_ptr<flutter::DlColorFilter> filter)
    : filter_(std::move(filter)) {}

ColorFilter::~ColorFilter() = default;

const std::shared_ptr<flutter::DlColorFilter>& ColorFilter::GetColorFilter()
    const {
  return filter_;
}

}  // namespace impeller::interop
